util/conditions/patch_test.go

```diff
@@ -0,0 +1,262 @@
+					Op:     RemoveConditionPatch,
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			got := GetPatch(tt.current, tt.target)
+
+			g.Expect(got).To(Equal(tt.want))
+		})
+	}
+}
+
+func TestApplyPatch(t *testing.T) {
+	fooTrue := TrueCondition("foo")
+	fooFalse := FalseCondition("foo", "reason foo", clusterv1.ConditionSeverityInfo, "message foo")
+	fooWarning := FalseCondition("foo", "reason foo", clusterv1.ConditionSeverityWarning, "message foo")
+
+	tests := []struct {
+		name    string
+		source  clusterv1.Conditions
+		patch   []ConditionPatch
+		want    clusterv1.Conditions
+		wantErr bool
+	}{
+		{
+			name:    "No patch return same list",
+			source:  conditionList(fooTrue),
+			patch:   []ConditionPatch{},
+			want:    conditionList(fooTrue),
+			wantErr: false,
+		},
+		{
+			name:   "Add: When a condition does not exists, it should add",
+			source: conditionList(),
+			patch: []ConditionPatch{
+				{
+					Base:   nil,
+					Target: fooTrue,
+					Op:     AddConditionPatch,
+				},
+			},
+			want:    conditionList(fooTrue),
+			wantErr: false,
+		},
+		{
+			name:   "Add: When a condition already exists but without conflicts, it should add",
+			source: conditionList(fooTrue),
+			patch: []ConditionPatch{
+				{
+					Base:   nil,
+					Target: fooTrue,
+					Op:     AddConditionPatch,
+				},
+			},
+			want:    conditionList(fooTrue),
+			wantErr: false,
+		},
+		{
+			name:   "Add: When a condition already exists but with conflicts, it should error",
+			source: conditionList(fooTrue),
+			patch: []ConditionPatch{
+				{
+					Base:   nil,
+					Target: fooFalse,
+					Op:     AddConditionPatch,
+				},
+			},
+			want:    nil,
+			wantErr: true,
+		},
+		{
+			name:   "Remove: When a condition was already deleted, it should pass",
+			source: conditionList(),
+			patch: []ConditionPatch{
+				{
+					Base:   fooTrue,
+					Target: nil,
+					Op:     RemoveConditionPatch,
+				},
+			},
+			want:    conditionList(),
+			wantErr: false,
+		},
+		{
+			name:   "Remove: When a condition already exists but without conflicts, it should delete",
+			source: conditionList(fooTrue),
+			patch: []ConditionPatch{
+				{
+					Base:   fooTrue,
+					Target: nil,
+					Op:     RemoveConditionPatch,
+				},
+			},
+			want:    conditionList(),
+			wantErr: false,
+		},
+		{
+			name:   "Remove: When a condition already exists but with conflicts, it should error",
+			source: conditionList(fooTrue),
+			patch: []ConditionPatch{
+				{
+					Base:   fooFalse,
+					Target: nil,
+					Op:     RemoveConditionPatch,
+				},
+			},
+			want:    nil,
+			wantErr: true,
+		},
+		{
+			name:   "Change: When a condition exists without conflicts, it should change",
+			source: conditionList(fooTrue),
+			patch: []ConditionPatch{
+				{
+					Base:   fooTrue,
+					Target: fooFalse,
+					Op:     ChangeConditionPatch,
+				},
+			},
+			want:    nil,
+			wantErr: true,
```

Comment:
should this not expect an error and the `want` to include `fooFalse`

Reply:
good catch, you are right!